File: java-client/src/main/java/org/scassandra/http/client/Query.java
/*
 * Copyright (C) 2014 Christopher Batey
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.scassandra.http.client;

public final class Query {
    
    public static class QueryBuilder {

        private String query;
        private String consistency = "ONE";

        private QueryBuilder() {}

        public QueryBuilder withQuery(String query){
            this.query = query;
            return this;
        }

        /**
         * Defaults to ONE if not set.
         * @param consistency Query consistency
         * @return this builder
         */
        public QueryBuilder withConsistency(String consistency){
            this.consistency = consistency;
            return this;
        }
        
        public Query build(){
            if (query == null) {
                throw new IllegalStateException("Must set query");
            }
            return new Query(this.query, this.consistency);
        }
    }

    public static QueryBuilder builder() { return new QueryBuilder(); }

    private final String query;
    private final String consistency;

    private Query(String query, String consistency) {
        this.query = query;
        this.consistency = consistency;
    }

    public String getQuery() {
        return query;
    }

    public String getConsistency() {
        return consistency;
    }

    @Override
    public String toString() {
        return "Query{" +
                "query='" + query + '\'' +
                ", consistency='" + consistency + '\'' +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Query query1 = (Query) o;

        if (consistency != null ? !consistency.equals(query1.consistency) : query1.consistency != null) return false;
        if (query != null ? !query.equals(query1.query) : query1.query != null) return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = query != null ? query.hashCode() : 0;
        result = 31 * result + (consistency != null ? consistency.hashCode() : 0);
        return result;
    }
}

File: codec/src/main/scala/org/scassandra/codec/messages/Query.scala
/*
 * Copyright (C) 2016 Christopher Batey and Dogan Narinc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.scassandra.codec.messages

import org.scassandra.codec.Consistency._
import org.scassandra.codec.Notations.{consistency, queryValue, bytes => cbytes, int => cint, long => clong, short => cshort}
import org.scassandra.codec.{Consistency, ProtocolVersion, QueryValue}
import scodec.Codec
import scodec.bits.ByteVector
import scodec.codecs._
import shapeless.{::, HNil}

import scala.{:: => ~~}

case class QueryParameters(
  consistency: Consistency = ONE,
  values: Option[List[QueryValue]] = None,
  skipMetadata: Boolean = false,
  pageSize: Option[Int] = None,
  pagingState: Option[ByteVector] = None,
  serialConsistency: Option[Consistency] = None,
  timestamp: Option[Long] = None
)

object DefaultQueryParameters extends QueryParameters()

object QueryParameters {
  private[this] lazy val v1FlagsCodec = provide(QueryFlags())

  implicit def codec(implicit protocolVersion: ProtocolVersion): Codec[QueryParameters] = {
    ("consistency"        | Consistency.codec)                                                                ::
    ("flags"              | withDefault(conditional(protocolVersion.version > 1, Codec[QueryFlags]), v1FlagsCodec)).consume { flags =>
    ("values"             | conditional(flags.values, listOfN(cshort, queryValue(flags.namesForValues))))     ::
    ("skipMetadata"       | provide[Boolean](flags.skipMetadata))                                             ::
    ("pageSize"           | conditional(flags.pageSize, cint))                                                ::
    ("pagingState"        | conditional(flags.withPagingState, cbytes))                                       ::
    ("serialConsistency " | conditional(flags.withSerialConsistency, consistency))                            ::
    ("timestamp"          | conditional(flags.withDefaultTimestamp, clong))
  } { data =>
      // This is admittedly contrived, but allows us to forgo making QueryFlags part of QueryParameters as
      // they should be derivable by the parameters themselves.
      val values :: skipMetadata :: pageSize :: pagingState :: serialConsistency :: timestamp :: HNil = data
      val valuesPresent = values.isDefined && values.getOrElse(Nil).nonEmpty
      val namesForValues = values match {
        case Some(h ~~ tail) => h.name.isDefined
        case _ => false
      }
      QueryFlags(namesForValues, timestamp.isDefined, serialConsistency.isDefined, pagingState.isDefined,
        pageSize.isDefined, skipMetadata, valuesPresent)
    }
  }.as[QueryParameters]
}

private[this] case class QueryFlags(
  namesForValues: Boolean = false,
  withDefaultTimestamp: Boolean = false,
  withSerialConsistency: Boolean = false,
  withPagingState: Boolean = false,
  pageSize: Boolean = false,
  skipMetadata: Boolean = false,
  values: Boolean = false
)

private[this] object QueryFlags {
  implicit val codec: Codec[QueryFlags] = {
    ("reserved"          | ignore(1)) ::
    ("namesForValues"    | bool)      ::
    ("defaultTimestamp"  | bool)      ::
    ("serialConsistency" | bool)      ::
    ("pagingState"       | bool)      ::
    ("pageSize"          | bool)      ::
    ("skipMetadata"      | bool)      ::
    ("values"            | bool)
  }.as[QueryFlags]
}